engine/src/witness/common/chunked_chain_source/latest_then.rs

```diff
@@ -19,6 +20,137 @@ impl<Inner, F> LatestThen<Inner, F> {
 		Self { inner, f }
 	}
 }
+
+fn latest_then_stream<'a, ItemStream, Fut, ThenFn, Output, Index, Hash, Data, Info, HistoricInfo>(
+	chain_stream: Pin<Box<ItemStream>>,
+	epoch: Epoch<Info, HistoricInfo>,
+	then_fn: &'a ThenFn,
+) -> BoxChainStream<'a, Index, Hash, Output>
+where
+	ItemStream: Stream<Item = Header<Index, Hash, Data>> + Send + 'a + ?Sized,
+	ThenFn:
+		Fn(Epoch<Info, HistoricInfo>, Header<Index, Hash, Data>) -> Fut + Send + Sync + Clone + 'a,
+	Fut: Future<Output = Output> + Send,
+	Output: Send + Sync + Unpin + 'static,
+	Index: aliases::Index,
+	Hash: aliases::Hash,
+	Data: aliases::Data,
+	Info: Clone + Send + Sync + 'static,
+	HistoricInfo: Clone + Send + Sync + 'static,
+{
+	stream::unfold(
+			(epoch.clone(), chain_stream, None),
+			move |(epoch, mut chain_stream, option_old_then_fut)| {
+				async move {
+					let apply_then = |header: Header<_, _, _>| {
+						println!("[MAXIM] apply then is called");
+						let epoch = epoch.clone();
+						#[allow(clippy::redundant_async_block)]
+						header
+							.then_data(move |header| async move { then_fn(epoch, header).await })
+							.boxed()
+					};
+
+					let (
+						// The future for the first header we see
+						mut option_first_then_fut,
+						// The future for the newest header we've seen
+						mut option_newest_then_fut,
+					) = {
+						// skip forward to newest ready item
+						let option_latest_fut = {
+							let mut option_latest = None;
+							loop {
+								match chain_stream.next().now_or_never() {
+									Some(None) => return None,
+									Some(Some(item)) => option_latest = Some(item),
+									None => break option_latest
+								}
+							}
+						}
+						.map(apply_then);
+
+						match option_old_then_fut {
+							Some(old_then_fut) => (Some(old_then_fut), option_latest_fut),
+							None => (option_latest_fut, None),
+						}
+					};
+
+					loop_select!(
+						if let Some(newest_header) = chain_stream.next() => {
+							*if option_first_then_fut.is_none() {
+								&mut option_first_then_fut
+							} else {
+								&mut option_newest_then_fut
+							} = Some(apply_then(newest_header));
+						} else break None,
+						if option_first_then_fut.is_some() => let mapped_header = option_first_then_fut.as_mut().unwrap() => {
+							// Keep replaceable_then_fut as it is from a newer header than persistent_then_fut
+							break Some((mapped_header, (epoch, chain_stream, option_newest_then_fut)))
+						},
+						if option_newest_then_fut.is_some() => let mapped_header = option_newest_then_fut.as_mut().unwrap() => {
+							// Don't keep persistent_then_fut as it is from an older header than replaceable_then_fut
+							break Some((mapped_header, (epoch, chain_stream, None)))
```

Comment:
I see that the first branch also wasn't executed, so I added a delay in the stream to test it too.

As for the last branch, I'm not sure if that code is even reachable. From reading the code (unfortunately the only existing comments seem to be out of date and refer to non-existing variables), it looks like that branch is triggered only when `option_first_then_fut` is None and `option_newest_then_fut` is Some. But `newest` is only set in two places: (1) in the first branch when `first` is Some (but that would trigger the second branch), or (2) in the code before loop_select where for `(first, newest)` we have either `(Some, option_latest_fut)` or `(option_latest_fut, None)`. Again we are looking for (None, Some) to enter the last branch in the end.

What external condition is it even meant to represent? Maybe I'm missing something, but it looks like it can be removed.

Reply:
So I just dove into it, I think you're missing the fact that the apply_then is also async. And we have to assume it can take arbitrarily long. Here's a test that hits the case, due to the non-determinism of select, it takes slightly differing paths each time it seems.

Reply:
(I would leave it as a separate test, maybe even separate the current one into 2, so it is a bit easier to reasona about the expected behaviour by looking at one particular property being tested by each test)

Reply:
I did try adding sleep in `apply_then` earlier and still didn't hit the third branch... I will take a look at your changes on Monday. Thanks!

Reply:
OK, I see how you tirggered the third branch, and I think I mangaged to reduce it to a minimal example that still behaves the same. All three branches are in separate tests now. (Sorry I didn't merge your PR directly. Was experimenting in my own branch, and was easiest to just push it.)